File: solutions/day02/solution.jl
# Define input path
input_path = (@__DIR__) * "/input"

function load(file_path=input_path)
    # Read data from file
    data = read((@__DIR__) * "/input", String)
    # Split into lines
    lines = split(strip(data), "\n")
    # Create named tuple from data
    return [
            (
            range    = parse.(Int, split(x[1], '-')),
            value    = x[2][1],
            password = x[3]
            )   for x in [split(x) for x in lines]]
end

function task1(data)
    """
    Return number of passwords that meet the condition:
        first <= occurances(value, password) <= second
    """
    return count(
        p -> p.range[1] <= count(x-> x == p.value, p.password) <= p.range[2],
        data)
end

function task2(data)
    """
    Return number of passwords that meet the condition:
        (password[first] == value) xor (password[second] == value)
    """
    return count(
        p -> (p.password[p.range[1]] == p.value) != (p.password[p.range[2]] == p.value),
        data)
end

function solution()
    data = load()
    println("First: $(task1(data))")
    println("Second: $(task2(data))")
end

@time solution()

File: solutions/day02/solution.py
policy = [(*map(int, x[0].split("-")), x[1][0], x[2]) for x in [x.split(" ") for x in open("input").read().strip().split('\n')]]
print(f"First: {len([1 for x in policy if x[0] <= x[3].count(x[2]) <= x[1]])}")
print(f"Second: {len([1 for x in policy if (x[3][x[0]-1] == x[2]) != (x[3][x[1]-1] == x[2])])}")

"""
# Parse inputs to tuple(first, second, 'value', 'password')
policy = [(*map(int, x[0].split("-")), x[1][0], x[2]) for x in [x.split(" ") for x in open("input").read().strip().split('\n')]]

# First star
print(f"First: {
    len(    # Get length of resulting array
        [1 for x in policy if # Add one to array if condition is valid
            x[0] <= x[3].count(x[2]) <= x[1] # Number of values in password is in range first <= count <= second
        ]
    )}")

# Same as first, but with a bit more complicated condition
# No zero indexing
print("Second: {}".format(
    len(                                        # Get length of resulting array
        [1 for x in policy if                   # Add one to array if condition is valid
            (x[3][x[0]-1] == x[2]) != (x[3][x[1]-1] == x[2])
        ]
    )))
"""